signalfx-wrapper.js

```diff
@@ -12,67 +14,99 @@ class SignalFxWrapper {
     this.originalEvent = originalEvent;
     this.originalContext = originalContext;
     this.originalCallback = originalCallback;
+    this.enableMetrics = sfxHelper.isMetricsEnabled(disableMetrics);
+    this.enableTracing = sfxHelper.isTracingEnabled(disableTracing);
 
-    sfxHelper.setAccessToken(accessToken);
-    sfxHelper.setLambdaFunctionContext(this.originalContext, dimensions);
-    sfxHelper.sendCounter('function.invocations', 1);
+    this.metricSender = sfxHelper;
+    if (!this.enableMetrics) {
+      this.metricSender = noopMetricSender;
+    }
+
+    this.execMeta = sfxHelper.getExecutionMetadata(this.originalContext);
+    sfxHelper.setDefaultDimensions(dimensions, this.execMeta);
+
+    this.metricSender.sendCounter("function.invocations", 1);
     if (coldStart) {
-      sfxHelper.sendCounter('function.cold_starts', 1);
+      this.metricSender.sendCounter("function.cold_starts", 1);
       coldStart = false;
     }
+
     return this;
   }
 
   invoke() {
     var exception, error, message, callbackProcessed;
 
+    const tracer = tracing.tracer(!this.enableTracing);
+    const span = tracing.startSpan(tracer, this.originalEvent, this.execMeta);
+
     const startTime = new Date().getTime();
 
     const processCallback = () => {
       if (callbackProcessed) {
         return;
       }
       callbackProcessed = true;
-      sfxHelper.sendGauge('function.duration', new Date().getTime() - startTime);
+      this.metricSender.sendGauge(
+        "function.duration",
+        new Date().getTime() - startTime
+      );
+
+      const err = exception || error;
+      if (err) {
+        span.addTags({
+          "sfx.error.kind": err.name,
+          "sfx.error.message": err.message,
+          "sfx.error.stack": err.stack,
+        });
+      }
 
       const runCallback = () => {
         if (exception) {
-          this.originalCallback(exception, 'Exception was thrown');
+          this.originalCallback(exception, "Exception was thrown");
         }
         this.originalCallback(error, message);
-      }
-      sfxHelper.waitForAllSends().then(runCallback, runCallback);
-    }
+      };
+
+      span.finish();
+
+      Promise.all([tracing.flush(), sfxHelper.waitForAllSends()]).then(
+        runCallback,
+        runCallback
+      );
+    };
 
     const customCallback = (err, msg) => {
       error = err;
       message = msg;
       processCallback();
-    }
+    };
 
-    try {
-      this.originalFn.call(
-        this.originalObj,
-        this.originalEvent,
-        this.originalContext,
-        customCallback
-      );
-    } catch (err) {
-      sfxHelper.sendCounter('function.errors', 1);
-      exception = err;
-      processCallback();
-    }
+    tracer.scope().activate(span, () => {
+      try {
+        this.originalFn.call(
+          this.originalObj,
+          this.originalEvent,
+          this.originalContext,
+          customCallback
+        );
+      } catch (err) {
```

Comment:
Is no longer incrementing the `function.errors` counter intentional?

Reply:
Fixed.